# Project: Exploring Car Sales Data from ebay

<b>This project was created to showcase my ability to clean / manipulate data and draw conclusions from the dataset using Pandas.</b>

In this project, I'll be working with a modified data set of used car listings from ebay's German website. The reason I picked this data set is because, I enjoy following the automotive industry, and secondly, this dataset was easily accessible online. <a href="https://data.world/data-society/used-cars-data">Dataset</a>

## Project Outline:
#### 1. Summary of the Data
- 1.1 <b>Data Dictionary</b>
A brief summary of the dataset and the attributes that make up the dataset.

#### 2. Uncovering unworkable Data Points
Exploring the data and looking for datapoints that may be null or need modifications to proceed with an analysis.
#### 3. Cleaning the Data
Cleaning the data by transforming datatypes of columns, or dropping columns.
#### 4. Analyzing the Data
Looking for interesting patterns and correlations within the dataset.
#### 5. Conclusion
Summarizing what insights have been found, and how knowing these insights would benefit a person buying a used car.

## 1. Summary of the Data
This dataset was originally scraped and uplaoded to Kaggle by <a href="https://www.kaggle.com/orgesleka">user:orgesleka</a>. Unfortunately, this user no longer exists within the Kaggle community. A modified version of the dataset was found on <a href="https://data.world/data-society/used-cars-data"> DataWorld </a>. The modified version is a sample of the original, where (50,000) rows were sampled from the original.

<b>Note:</b> Since this data was scraped there will be plenty of things to clean up before a proper analysis can begin. Once everything is cleaned up, we'll explore, analyze, and call out any useful / actionable patterns.  

### 1.1 Data Dictionary - Summary of each column
<b> Please refer to this if any column confuses you.</b>
- `dateCrawled` - When the ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

## 2. Uncovering Unworkable Data Points
Within this section, we'll explore and look for data points that may be null or missing. We'll also look for data that doesn't match within a column. 

In [1]:
#importing libraries
import numpy as np
import pandas as pd

#reading in data
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
#exploring data set using summary methods
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

This dataset is made up of 20 columns and 50K rows. Most of the columns are stored as strings. It looks like there are a few columns with NULL values. There are some columns that also have dates stored as strings. 

### 2.1 Cleaning Column Names
Our columns are all in camelCase, I'm going to change them to snakecase to make them easier to look at. 
Some of the columns are also a little vague, I'll rename them with a more descriptive title.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

#### Transforming the columns into snakecase.

In [4]:
#using a dictionary to hold new column names
new_name_dict = {
    "yearOfRegistration": "registration_year", 
    "monthOfRegistration": "registration_month",
    "notRepairedDamage": "unrepaired_damage",
    "dateCreated": "ad_created",
    "dateCrawled": "date_crawled",
    "offerType": "offer_type",
    "vehicleType": "vehicle_type",
    "powerPS": "power_ps",
    "fuelType": "fuel_type",
    "nrOfPictures": "num_of_pics",
    "postalCode": "postal_code",
    "lastSeen": "last_seen"}

#converting column names from camelCase to snakecase.
autos.rename(new_name_dict, axis = 1, inplace = True)
#checking new column names
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_of_pics', 'postal_code',
       'last_seen'],
      dtype='object')

### 2.2 Exploring the Data
In this section we'll start exploring the data and looking for obvious areas where we can begin cleaning data.

In [5]:
#looking at descriptive stats for all columns
autos.describe(include = "all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_of_pics   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

#### Initial Observations
There are three columns (seller, offer_type, _num_of_pics) that have values that are nearly the same for each row. 
Out of all the columns, there are three that need further analysis: 
- `seller` 
- `offer_type` 
- `num_of_pics`.
<br>
Most values in these columns look the same.
These columns will be analyzed further below.

In [6]:
# num_of_pics column values
autos["num_of_pics"].value_counts()

0    50000
Name: num_of_pics, dtype: int64

Within this column, there looks to be a 0 for every column. This column will be dropped. 

In [7]:
#Checking for values within seller column
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [8]:
#inspecting offer_type column further
autos.head()
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

After gathering the values within the "seller" and "offer_type" columns, they seem to also hold fixed values that would be irrelevant to our analysis. These columns will be dropped too.

### 3. Cleaning the Data
Within this secton, we'll be dropping columns and cleaning up data from other columns. 

In [9]:
#dropping columns
autos = autos.drop(["seller", "num_of_pics","offer_type"], axis = 1)

In [10]:
#checking columns
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

### 3.1 Removing non-integer data from price column and observations
In the price column, we've found several values containing a comma and a dollar sign. We'll remove the comma and dollar sign to make all values uniform. The values seem to be pretty uniform with a couple of outliers that are not rounded off. Perhaps eBays system automatically rounded off values the users input. 

In [12]:
autos['price'].value_counts()

$0            1421
$500           781
$1,500         734
$2,500         643
$1,000         639
              ... 
$14,998          1
$55,500          1
$3,890,000       1
$37,400          1
$2,910           1
Name: price, Length: 2357, dtype: int64

In [13]:
#remove non integer values from price column
autos["price"] = (autos["price"]
                  .str.replace("$","")
                  .str.replace(",","")
                  .astype(int))

#### 3.2 Removing non-integer values from odometer column and observations
From below we can see that the `odometer` column contains non-integer values of [,] and [km]. These will be removed to make the column values uniform. Note that there are also more high mileage cars versus low mileage cars. Values also seem to be rounded off indicating that users had to select from a drop down.

In [15]:
autos["odometer"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [16]:
#remove non int values from odometer column
autos["odometer"] = (autos["odometer"]
.str.replace("km","")
.str.replace(",","")
.astype(int))

In [28]:
# Checking data type of columns after formatting
print(autos["price"].dtypes)
print(autos["odometer"].dtypes)

int64
int64


Because Germany uses kilometers instead of miles, we'll rename the column odometer_km to give more description to the column.

In [29]:
#renaming odometer column to odometer_km
autos.rename({"odometer": "odometer_km"}, axis = 1, inplace = True)

In [30]:
#checking all column names and dtypes
print(autos.dtypes)

date_crawled          object
name                  object
price                  int64
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer_km            int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
postal_code            int64
last_seen             object
dtype: object


## Part 4: Exploring Odometer and Price columns

In [ ]:
autos["odometer_km"].value_counts()

From here we can see that there are more high mileage vehicles, than there are with low mileage.

Within the odometer column, values seem to be rounded off. There are no values that are odd or oddly numbered. They all seem to fit some convention. Perhaps ebay only allows mileage to be input as a rounded number. Or the API automatically rounds up or down. There doesn't seem to be any outliers that would affect our analysis.

In [31]:
#checking how many unique values in price column
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

In [32]:
#further inspecting price column
autos["price"].value_counts().sort_index().tail(20)

197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price, dtype: int64

In [33]:
autos["price"].value_counts().sort_index().head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

Just as the values in the odometer column were rounded, the same seems to apply to the price column. There are also many values which are below zero(1421) and some that are extremely high for being on ebay. Although these values may be real, we want to keep our analysis specifically to values that fit within normal ranges.

I will remove prices up to a 350K dollars since there may be cars of great value that have sold online.

In [35]:
#removing values from within price column
autos = autos[autos["price"].between(1,351000)]

In [36]:
autos["price"].describe().round(2)

count     48565.00
mean       5888.94
std        9059.85
min           1.00
25%        1200.00
50%        3000.00
75%        7490.00
max      350000.00
Name: price, dtype: float64

Now we can work with values with a price range of (0, 350,000).

### 4.2 Exploring the Date Columns

In [45]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

#### Observations on Date Data 
Looking at our dataset, there are three columns that follow a certain date format, those being 
- `date_crawled` 
- `last_seen` 
- `ad_created` 
<br>These columns were formatted this way due to a crawler that automatically captured the data for these columns. The other two date columns (`registration_month`, `registration_year`) seem to use a single value for either month or year.

In [46]:
#Checking formatting of three string columns
autos[["date_crawled", "ad_created", "last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [47]:
(autos["date_crawled"]
.str[:10]
.value_counts(normalize = True, dropna = False)
.sort_index() #sorts based on first column cells
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [48]:
(autos["date_crawled"]
.str[:10]
.value_counts(normalize = True, dropna = False)
.sort_values(ascending = True) #sorts based on values not index col 2
)

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

From this data we can see that the site was crawled over the course of a month from March 2016 to April 2016. There were a couple of days the when it seems like the crawler had some downtime. For the most part most of the data seems level.

#### Working on last_seen column
These values indicate when the listing was last crawled which can help us determine when a listing was removed. During the last three days there seems to be a spike compared to other days. It's unlikely however that this is the case. In the date_crawled column for these same dates, the crawler exhibited less activity, indicating that the crawler just didn't collect as many data points compared to previous days.

In [49]:
(autos["last_seen"]
 .str[:10]
 .value_counts(normalize = True, dropna = False)
 .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

#### Working on ad_created Column
Within this column we can see when an ad was created. It looks like Ebay keeps listings on a yearly basis. Most of the data points with significance fall within March to April. There are however some older listings going back nine months.

In [50]:
(autos["ad_created"]
.str[:10]
.value_counts(normalize = True, dropna = False)
.sort_index())

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

### 4.3 Analyzing the registration_year and registration_month Columns
Compared to the date columns we worked with above, these two only exhibit a single date value. We will explore and clean up any uncessary characters in these columns

In [51]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [52]:
autos["registration_year"].value_counts().sort_values(ascending = False)

2000    3156
2005    2936
1999    2897
2004    2703
2003    2699
        ... 
1953       1
4500       1
1939       1
1001       1
1952       1
Name: registration_year, Length: 95, dtype: int64

Immediately we can see a problem with the registration_year column. There are values that shouldn't exist such as the year(s) 1000 and 9999. It's impossible to have cars from these years. Becasue it's Ebay and an auction site there may be some very antique cars to bid on. As such I'm going to include only cars from (1900-2016).

### 4.4 Dealing with Incorrect Registration Year Data


In [53]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Using the sum function, we can assess what part of the data of our specified range (1900 - 2016), makes up the dataset. If the value if insignificant, then we'll have to adjust our interval.

using a "~" sign reveals the opposite result of our calculation, it says exclude these values. Without the ~ sign we 
include the values

In [54]:
#including only cars from 1900 - 2016
(autos["registration_year"].between(1900, 2016)).sum() / autos.shape[0]


0.961206630289303

96% of the datapoints are contained in the interval (1900, 2016). All the remaining data points outside of our interval will be excluded. So our sample is the data with registration_year of cars being from (1900 - 2016). Any significant discoveries can be made with 95% of the data having been analyzed. Therefore these rows will be removed.

In [55]:
#Removing the rows outside our interval
autos = autos[autos["registration_year"].between(1900,2016)]

In [56]:
autos.shape

(46681, 17)

## Changes made to dataset
We went from 50000 rows and 20 columns to 46681 rows and 17 columns. This will allow us to work with data that provides a great starting point for analysis.

In [ ]:
#value counts for series 
autos["registration_year"].value_counts(normalize = True).sort_values(ascending = False)

From our analysis above it looks like most cars were registered within 20 years of 2016. 

## Exploring Price by Brand

Finding out which brand pulls in the most sales, is a critical part of this analysis. Once we know which brands are selling we can make better predictions on what types of vehicles people want and the trends that have been shown.

In [57]:
#finding each brand of car
sort_autos = autos["brand"].unique()
print(sorted(sort_autos))

['alfa_romeo', 'audi', 'bmw', 'chevrolet', 'chrysler', 'citroen', 'dacia', 'daewoo', 'daihatsu', 'fiat', 'ford', 'honda', 'hyundai', 'jaguar', 'jeep', 'kia', 'lada', 'lancia', 'land_rover', 'mazda', 'mercedes_benz', 'mini', 'mitsubishi', 'nissan', 'opel', 'peugeot', 'porsche', 'renault', 'rover', 'saab', 'seat', 'skoda', 'smart', 'sonstige_autos', 'subaru', 'suzuki', 'toyota', 'trabant', 'volkswagen', 'volvo']


In [58]:
#finding which brands have the most listings
(autos["brand"].value_counts(normalize = True))

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

German auto brands seem to dominate the number of listings on Ebay. This is most likely attributed to the data being pulled from Ebay for Germany. We can factor that into our analysis. Volkswagen is the dominant brand accompanying for over 20% of listings. <b>Any company that occupies less than 5% of the listing space will be excluded as they will not be massive influcences on our analysis.</b>

### Aggregating data by brand

In [61]:
#creating counts for all brands
num_brands = autos["brand"].value_counts(normalize = True)
#only selecting brands that make up > .05 percent of listings as an index
large_brands = num_brands[num_brands > .05].index
print(large_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [62]:
large_brands.shape

(6,)

After gathering brands that were less than 5% in listings, we've found that 6 are the most common on the site.

In [63]:
#Aggregating brand avg sell price
avg_brand_price = {}

for brand in large_brands:
    only_brand = autos[autos["brand"] == brand] #only_brand is equal to only rows that match brand name in loop.
    brand_avg = only_brand["price"].mean() #find the avg of all rows for the brand being executed.
    avg_brand_price[brand] = int(brand_avg) #add the values to the dictionary and convert price to int.
    

avg_brand_price

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

### The Results of Price related to brand
Here is the affordability of brands in order: audi, mercedes_benz, bmw, volkswagen, ford, and opel. <br>Although BMW, Audi, and Mercedez-Benz are all higher than our other three brands(ford, Volkswagen, Opel), they still have a great amount of listings. 

Ford and Opel are significantly less expensive than the 3 luxury brands, however Volkswagen seems to be a middle ground in terms of pricing. Perhaps this can be attributed to Volkswagen being a major brand throughout Germany as they were a pioneer in the auto industry early on. Also because the most expensive brands are being resold it goes to say that they will have a higher resale value as well.

### Exploring Mileage

In [65]:
mean_prices_series = pd.Series(avg_brand_price)
pd.DataFrame(mean_prices_series, columns = ["mean_price"])

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [66]:
#aggregating avg mileage
avg_km = {}

for brand in large_brands:
    only_brand = autos[autos["brand"] == brand]
    km_brand = only_brand["odometer_km"].mean()
    avg_km[brand] = int(km_brand)
    
    
mean_prices = pd.Series(avg_brand_price).sort_values(ascending=False)
mean_km = pd.Series(avg_km).sort_values(ascending=False)

In [67]:
brand_info = pd.DataFrame(mean_km, columns = ["mean_mileage"])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [68]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

In [69]:
brand_info = brand_info.reindex(columns = ["mean_price", "mean_mileage"])
brand_info

mean_price  mean_mileage
bmw                  8332        132572
mercedes_benz        8628        130788
opel                 2975        129310
audi                 9336        129157
volkswagen           5402        128707
ford                 3749        124266

Luxury brands seem to have no benefit in terms of lower mileage for higher price. The other brands(ford, Opel, and Volkswagen) show a greater correlation between mileage and price but this is not completely true for all brands. Expensive vehicles have more mileage whereas less expensive vehicles have lower mileage.

In [76]:
brand_info.corr()

mean_price  mean_mileage
mean_price      1.000000      0.612335
mean_mileage    0.612335      1.000000

## 5. Conclusion

#### From our analysis above we can a positive correlation between price and mileage. This indicates that a higher price may indicate a higher mileage amount. This seems to be the case for luxury brands and is the opposite case for the non-luxury brands (Ford, Opel, and Volkswagen).